src/ashley/views.py

```diff
@@ -0,0 +1,117 @@
+"""Views of the ashley django application."""
+
+import logging
+from abc import ABC, abstractmethod
+from typing import List
+
+from django.contrib.auth import authenticate, login
+from django.http import (
+    HttpResponse,
+    HttpResponseBadRequest,
+    HttpResponseForbidden,
+    HttpResponseRedirect,
+)
+from django.urls import reverse
+from django.utils import translation
+from django.utils.decorators import method_decorator
+from django.views import View
+from django.views.decorators.csrf import csrf_exempt
+from machina.apps.forum_permission.shortcuts import assign_perm
+from machina.core.db.models import get_model
+
+from lti_provider.exceptions import LTIException
+from lti_provider.lti import LTI
+
+from .settings import settings
+
+Forum = get_model("forum", "Forum")  # pylint: disable=C0103
+LTIContext = get_model("ashley", "LTIContext")  # pylint: disable=C0103
+
+logger = logging.getLogger("ashley")
+
+
+@method_decorator(csrf_exempt, name="dispatch")
+class BaseLTIView(ABC, View):
+    """Forum view called by an LTI launch request."""
+
+    def post(self, request, *args, **kwargs) -> HttpResponse:  # pylint: disable=W0613
+        """Handler for POST requests."""
+        lti_request = LTI(request)
+        try:
+            lti_request.verify()
+            user = authenticate(request, lti_request=lti_request)
+            if user is not None:
+                login(request, user)
+                return self._do_on_login(lti_request)
+            return HttpResponseForbidden("Forbidden")
+        except LTIException:
+            return HttpResponseForbidden("Forbidden")
+
+    @abstractmethod
+    def _do_on_login(self, lti_request: LTI) -> HttpResponse:
+        """Process the request when the user is logged in via LTI"""
+
+
+class ForumLTIView(BaseLTIView):
+    """Forum view called by an LTI launch request."""
+
+    def _do_on_login(self, lti_request: LTI) -> HttpResponse:
+        """Process the request when the user is logged in via LTI"""
+
+        # Get or create the LTIContext model associated with the current LTI launch request
+        context_id = lti_request.get_param("context_id")
+        if not context_id:
+            return HttpResponseBadRequest()
+        context, _context_created = LTIContext.objects.get_or_create(
+            remote_id=context_id, lti_consumer=lti_request.get_consumer()
+        )
+
+        # Synchronize the user groups related to the current LTI context
+        context.sync_user_groups(self.request.user, lti_request.roles)
```

Comment:
I think you didn't test the case of user that had more groups before the sync ie confirm that the extra group is removed?

Reply:
This case is tested here : 
https://github.com/openfun/ashley/blob/feature/unique-lti-link/tests/ashley/test_models_lticontext.py#L93

The user in the context2 has previously roles `role1` and `role2`.
After the sync, it is not a member of these groups anymore.